File: engine/schemas/js/src/canonical.mjs
import { createHash } from "node:crypto";
import canonicalize from "canonicalize";

export const CANONICALIZATION_ID = "urn:ietf:rfc:8785";
export const HASH_ALGORITHM = "sha-256";

export function canonicalText(value) {
  const text = canonicalize(value);
  if (text === undefined) {
    throw new TypeError("value cannot be serialized by RFC 8785/JCS");
  }
  return text;
}

export function canonicalBytes(value) {
  return Buffer.from(canonicalText(value), "utf8");
}

export function canonicalSha256(value) {
  return createHash("sha256").update(canonicalBytes(value)).digest("hex");
}

File: engine/compiler/src/nwe_compiler/canonical.py
from __future__ import annotations

import hashlib
from typing import Any

import rfc8785

CANONICALIZATION_ID = "urn:ietf:rfc:8785"
HASH_ALGORITHM = "sha-256"


def canonical_bytes(value: Any) -> bytes:
    """Serialize a JSON-compatible value using RFC 8785 / JCS."""
    return rfc8785.dumps(value)


def canonical_sha256(value: Any) -> str:
    """Return SHA-256 over RFC 8785 canonical bytes."""
    return hashlib.sha256(canonical_bytes(value)).hexdigest()
